File: Python_Scripts/Clustering/hierachical_clustering.py
# Hierarchical Clustering

# Importing Libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Setting working directory
import os
path = "/Users/takanaoishimura/Desktop/Udemy/Machine Learning A-Z Template Folder" \
    "/Part 4 - Clustering/Section 25 - Hierarchical Clustering"
os.chdir(path)
print(os.getcwd())
del(path)

# Importing the dataset
dataset = pd.read_csv("Mall_Customers.csv")
X = dataset.iloc[:, [3,4]].values

# Using the dendogram to find the optimal number of clusters
import scipy.cluster.hierarchy as sch
dendogram = sch.dendrogram(sch.linkage(X, method = "ward"))
plt.title("Dendogram")
plt.xlabel("Customers")
plt.ylabel("Euclidean Distance")
plt.show()

# Fit hierachical clustering to the dataset
from sklearn.cluster import AgglomerativeClustering
hc = AgglomerativeClustering(n_clusters = 5, 
                             affinity = "euclidean",
                             linkage = "ward")
y_hc = hc.fit_predict(X)

# Visualizing the clusters
colors = ['red', 'blue', 'green', 'cyan', 'magenta']
for i in range(0, 5):
    plt.scatter(X[y_hc == i, 0], 
                X[y_hc == i, 1], 
                c = colors[i], 
                label = 'Cluster ' + str(i+1))
plt.title('Clusters of Clients')
plt.xlabel('Annual Income (k$)')
plt.ylabel('Spending Score (1-100)')
plt.legend()
plt.show()

File: R_Scripts/Clustering/hierachical_clustering.R
# Hierachical Clustering

# Setting working directory
setwd(paste("/Users/takanaoishimura/Desktop/Udemy/Machine Learning A-Z Template Folder/",
            "Part 4 - Clustering/Section 25 - Hierarchical Clustering", sep = ""))

# Importing data
dataset = read.csv("Mall_Customers.csv")
X = dataset[4:5]

# Using the dendogram to find the optimal number of clusters
dendogram = hclust(d = dist(X, method = "euclidean"),
                   method = 'ward.D')
plot(dendogram,
     main = "Dendogram",
     xlab = "Customers",
     ylab = "Euclidean Distances")

# Fitting the Heirachical Clustering to the dataset
hc = hclust(d = dist(X, method = "euclidean"),
            method = 'ward.D')
y_hc = cutree(tree = hc,
              k = 5)

# Visualizing the clusters
library(cluster)
clusplot(x = X, 
         y_hc,
         lines = 0,
         shade = TRUE,
         color = TRUE, 
         labels = 4, 
         plotchar = FALSE,
         span = TRUE,
         main = paste("Clusters of Clients"),
         xlab = "Annual Income",
         ylab = "Spending Score")


